### DISCUSSION TITLE
Where is the code for "docker volume create"?

### DISCUSSION THREAD

**Discussion by wsw70** [moby/moby#48267] [Q&A] [answered]
I would like to have a look at the code that actually processes a `docker volume create` (not https://github.com/moby/moby/blob/master/client/volume_create.go which just makes a call to an API). 

I do not know much about the deeper internals of docker, but I would like to understand what the creation of a volume actually means from a filesystem and metadata perspective (to authoritatively answer [my question at SU](https://superuser.com/questions/1850650/how-to-import-into-docker-recovered-volumes))

**Comment by thaJeztah** [comment, marked as answer]
The easiest way to find code related to API endpoints is to follow the code from the API router; start with the `api/server/router` code for the endpoint you're looking for. In this case, one of the `volume` endpoints; https://github.com/moby/moby/blob/89bca301be72080cd8c948f76ff7818b85268975/api/server/router/volume/volume.go#L27-L40

The router can be found in the same package; https://github.com/moby/moby/blob/89bca301be72080cd8c948f76ff7818b85268975/api/server/router/volume/volume_routes.go#L90-L93

Which calls the `backend.Create` https://github.com/moby/moby/blob/89bca301be72080cd8c948f76ff7818b85268975/api/server/router/volume/volume_routes.go#L118-L124

The backend is an interface that's defined in https://github.com/moby/moby/blob/89bca301be72080cd8c948f76ff7818b85268975/api/server/router/volume/backend.go#L13-L19

In our case the interface is implemented by the "volume" service; https://github.com/moby/moby/blob/89bca301be72080cd8c948f76ff7818b85268975/volume/service/service.go#L65-L84

Which uses the volume store; https://github.com/moby/moby/blob/89bca301be72080cd8c948f76ff7818b85268975/volume/service/store.go#L477-L485

**Reply by wsw70** [reply]
Thanks a lot for the detailed answer. I will dig deeper in `create` to get to the actual filesystem operations.
